field.go

```diff
@@ -418,132 +418,311 @@ func Inline(val zapcore.ObjectMarshaler) Field {
 // them. To minimize surprises, []byte values are treated as binary blobs, byte
 // values are treated as uint8, and runes are always treated as integers.
 func Any(key string, value interface{}) Field {
+	// To work around go compiler assigning unreasonably large space on stack
+	// (4kb, one `Field` per arm of the switch statement) which can trigger
+	// performance degradation if `Any` is used in a brand new goroutine.
+	var (
+		t     zapcore.FieldType
+		i     int64
+		s     string
+		iface any
+	)
 	switch val := value.(type) {
 	case zapcore.ObjectMarshaler:
-		return Object(key, val)
+		t = zapcore.ObjectMarshalerType
+		iface = val
 	case zapcore.ArrayMarshaler:
-		return Array(key, val)
+		t = zapcore.ArrayMarshalerType
+		iface = val
 	case bool:
-		return Bool(key, val)
+		var ival int64
+		if val {
+			ival = 1
+		}
+		t = zapcore.BoolType
+		i = ival
 	case *bool:
-		return Boolp(key, val)
+		if val == nil {
+			t = zapcore.ReflectType
+			break
```

Comment:
I also considered pulling all the *p types out of the switch into a separate switch, like

```
	switch val := value.(type) {
	case *int, *bool, *int64, ...:
		if val == nil {
			t = zapcore.ReflectType
			return Field ...
		}
```
but that doesn't work, each of the types would need it's own branch, like

```
	switch val := value.(type) {
	case *int:
		if val == nil {
			t = zapcore.ReflectType
			return Field ...
	case *int64:
		if val == nil {
			t = zapcore.ReflectType
			return Field ...
		}
```
so each of the `*` cases would need to be duplicated in both switches.

Slight benefit in reduction of `if/else` nil pointer checks, but increase in the total function length. Not sure.